posts/2024-07-26-actuated-for-gitlab.md

```diff
@@ -0,0 +1,219 @@
+        --license-file ~/.openfaas/LICENSE \
+        --operator \
+        --clusterrole \
+        --jetstream \
+        --autoscaler
+      kubectl get secret -n openfaas
+      echo $OPENFAAS_LICENSE | wc
+    - kubectl create secret docker-registry gitlab-ci-pull \
+        --docker-server=$CI_REGISTRY \
+        --docker-username=$CI_REGISTRY_USER \
+        --docker-password=$CI_JOB_TOKEN \
+        --docker-email=docker@example.com \
+        --namespace openfaas-fn
+    - kubectl rollout status -n openfaas deploy/gateway --timeout=60s
+    - |
+      kubectl port-forward -n openfaas svc/gateway 8080:8080 &>/dev/null &
+      echo -n "$!" > kubectl-pid.txt
+
+      echo PID for port-fowarding: $(cat kubectl-pid.txt)
+
+    - faas-cli ready --attempts 120
+
+    - |
+      kubectl patch serviceaccount \
+        -n openfaas-fn default \
+        -p '{"imagePullSecrets": [{"name": "gitlab-ci-pull"}]}'
+
+    # Deploy the function that we build and pushed in the previous stage.
+    - |
+      echo $(kubectl get secret \
+        -n openfaas basic-auth \
+        -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo) |
+      faas-cli login --username admin --password-stdin
+    - faas-cli template pull stack
+    - CI_REGISTRY=$CI_REGISTRY CI_COMMIT_SHORT_SHA=$CI_COMMIT_SHORT_SHA faas-cli deploy
+
+    # Test the OpenFaaS function by invoking it.
+    - faas-cli ready bcrypt
+    - curl -i http://127.0.0.1:8080/function/bcrypt -d ""
+
+    - kill -9 "$(cat kubectl-pid.txt)"
+
+  tags:
+    - actuated-4cpu-8gb
+    - shell
+```
+
+## Private peering
+
+Actuated allows you to host your runners wherever you want, in the public cloud or on premise in your own data center. This means that in some cases it can be difficult to expose the Actuated agent when your servers are behind a corporate firewall or have no inbound internet access.
+
+One of our enterprise customers was running into this issue. To allow them to quickly deploy agents in different networks within their organisations we released private peering for the agent. When peering is enabled the client makes an outbound connection to the Actuated control plane and does not rely on any inbound data path.
+
+Agent peering simplifies agent setup and improves security:
+
+- No additional setup, firewall or routing rule changes required.
+- The agent is exposed privately to the Actuated control plane.
+- All traffic between the agent and the control plane is encrypted.
+
+![Private peering diagram](/images/2024-07-gitlab/peering.png)
+
+> Peering example for an enterprise with two agents within their own private firewalls.
+
+## Wrapping up
+
+We highlighted some of the features and benefits of running GitLab CI jobs on Actuated runners:
+
+- Secure isolated builds in ephemeral microVMs.
+- Fixed cost & Less management.
+- Easy on demand mixed use of Docker and Shell executors.
+- Optimised usage of available runners.
+- Easy setup in complex network environments.
+
```

Comment:
Compared to GitLab CI's built-in solution, there is Kubernetes cluster required. VMs are packed into a fleet of servers efficiently or kept in a queue until capacity becomes available.

Docker can be used natively without any of the risks with the built-in Kubernetes runners which use privileged Pods.

There is no need to add side-cars for BuildKit, Kaniko or to have to fight with the issues surrounding user-namespaces.

Everything is tested by our team, so you don't have to fine-tune or configure a Kernel, we ship the OS image and Kernel together and update them remotely.